doc/developer/design/20240625_subsource_deprecation_blue_green_sources.md

```diff
@@ -0,0 +1,318 @@
+
+We will allow more than one table to reference the same `<upstream_ref>`, using a
+potentially different set of columns. This allows a user to handle an upstream
+schema change by creating a new table for the same upstream table and then performing
+a blue-green swap operation to switch their downstream dependencies to the new table,
+and then drop the old one.
+
+The existing options to `CREATE SOURCE` for Postgres & MySQL sources that specify
+which tables to ingest (`FOR ALL TABLES`, `FOR SCHEMAS` and `FOR TABLES`) will be
+removed. Instead, all upstream tables to ingest from this source will need to be
+explicitly referenced in a `CREATE TABLE` statement. While this may seem controversial,
+more often than not these options cause upstream data that does not need to be
+ingested into Materialize to be so, and by using explicit statements for each
+table to be ingested it makes Materialize configuration much more amenable to
+object<->state mappings in tools like DBT and Terraform.
+
+We can eventually reintroduce syntactic sugar to perform a similar function to
+`FOR ALL TABLES` and `FOR SCHEMAS` if necessary.
+
+
+### Implementation Plan
+
+1. Separate the planning of *sources* and *subsources* such that sources are fully
+   planned before any linked subsources. Part of this work has been prototyped in:
+   https://github.com/MaterializeInc/materialize/pull/27320
+
+2. Introduce the concept of *read-only tables* that link to a source ingestion
+   and an upstream 'reference' specifying which upstream table to ingest into this table.
+
+3. Update the CREATE TABLE statement to allow creation of a read-only table
+   with a reference to a source and an upstream-reference:
+   (`CREATE TABLE <new_table> FROM SOURCE <source> (upstream_reference)`)
+
+4. Update planning for `CREATE TABLE` to include a purification step akin to the
+   purification for `ALTER SOURCE .. ADD SUBSOURCE`. This will verify the upstream
+   permissions, schema, etc for the newly added source-fed table.
+
+5. Update the underlying *source* (ingestion) and source_export/ingestion_export
+   structures to allow more than one *source_export* to refer to an upstream
+   reference, and each *source_export* to have its own projection of the upstream data
+
+5. Update the storage controller to use both *subsources* and *read-only tables*
+   as *source_exports* for existing multi-output sources (postgres & mysql)
+
+6. Update the source rendering operators to handle the new structures and allow
+   outputting the same upstream table to more than one souce export.
+
+7. Migrate existing sources to the new source model (make all sources 'multi-output' sources):
+   - For existing sources where the primary relation has the data (e.g. Kafka sources):
+     - Create a read-only table for the current primary relation called `<source>`
+     - Rename the source itself `<source>_progress` and drop the existing
+      `<source>_progress` object
```

Comment:
What about downstream views that reference the progress object? (Maybe there just aren't any! We can check production.)

Reply:
Ah yes -- the intent of the rename was to preserve those references. Maybe I should reword this to imply that we would attempt to preserve the existing `_progress` collection, but we would essentially point the top-level source statement at it directly and rename the source-name within that statement at the same time.

Then we would actually drop the collection currently associated with the top-level source statement.
However I may be missing some unintended side effects related to the errors that may be part of that source collection -- need to investigate what would happen to those